src/t_hash.c

```diff
@@ -497,17 +497,18 @@ void hashTypeConvertZiplist(robj *o, int enc) {
             unsigned int vlen;
             long long lval;
             char longstr[32] = {0};
-            unsigned int maxlen = 0;
-            unsigned char *lp = lpNew(0);
+            unsigned char *lp; 
 
+            /* Use ziplist's size to pre-allocate listpack,
+             * avoid realloc when lpPushTail. */
+            lp = lpNew(ziplistBlobLen(o->ptr)); 
```

Comment:
Indeed，I missing it.

Reply:
```compact``` felling is not necessary.
The memory is already shrunk in ```lpInsert```, and since we allocate a sufficient amount of memory at the beginning, ```realloc``` does not change the listpack memory address until the realloced memory is larger than ziplsit size.

Reply:
i'm not sure i follow you.
if lpInsert would compact, then making an initial big allocation is not helpful, since the first insertion will shrink it back. and if it doesn't then in case the original allocation was too big, we're left with that excess at the end of filling.

Reply:
Yes, the first insertion will shrink it back, when the next insert, there is still a large chunk of memory behind this chunk of memory that is not used, so when it is inserted again, only the memory size of the ```listpack``` is changed, and no memory copy is made, which is a performance improvement.

Reply:
I'll write a test on this code(to see if the memory address changes), to see if it works in both malloc and jemalloc.

Reply:
Maybe write some underlying code not using `malloc` will be quicker.
But the code will be ugly

Reply:
@oranagra After testing, the pre-allocated space, shrink and then expand, the address will not change.
#6281 This modification will also have the same problem, lpAppendInteger will also shrink.
I don't think it makes sense to shrink in lpInsert, if ```listpack``` is not pre-allocated, shrink is nop, if memory is pre-allocated, user should execute lpShrinkToFit manually.

Reply:
@sundb one of us is missing something.
In #6281 there's a change in `lpInsert` to call realloc only if we need to increase the allocation size (see the call to lp_malloc_size), but not when it wants to shrink it.
In the test you conducted above, it looks like you're adding about 60 bytes, on a pre-allocation of 100, so i think the realloc will be skipped. either that, or maybe you're not using jemalloc (libc malloc can return the same pointer even if it changes the allocation size).

in that case, if we made an allocation of 100 bytes, and used only 60, we do need a realloc to shrink it (which in jemalloc will certainly allocate a new address and do a memcpy).

If i'm missing something, and lpInsert does call realloc to shrink, then (at least with jemalloc) it'll allocate a new block and move the memory to a different pointer, eliminating the optimization of pre-allocation.

Reply:
Sorry for my lack of seriousness.
The listpack shrinks only when elements are removed from the listpack, or when the length becomes smaller after replacing the element.

Reply:
ok. so in this code (converting from ziplist to listpack) we only add new elements, so we never shrink, right?
in that case we do need to call `lpShrinkToFit` at the end.

Reply:
Yes.